### DISCUSSION TITLE
AD

### DISCUSSION THREAD

**Discussion by jbcaillau** [control-toolbox/control-toolbox.github.io#53] [Ideas]
[AD tuto](https://gdalle.github.io/AutodiffTutorial)

**Comment by jbcaillau** [comment]
[The elements of differentiable programming](https://arxiv.org/abs/2403.14606)

**Comment by jbcaillau** [comment]
@PierreMartinon @ocots @joseph-gergaud check this on sparse AD (thanks @amontoison)
https://adrianhill.de/iclr2025-sparse-autodiff/blog/sparse-autodiff/

**Reply by ocots** [reply]
Thanks. 

Quarto :-)

**Comment by amontoison** [comment]
@PierreMartinon @jbcaillau,  
I wanted to message you for a few days about coloring.  
I discovered that, depending on the initial ordering used during coloring, the number of colors can vary significantly.  
Since you almost always have the same sparsity pattern (due to your discretization scheme), it might be useful to try a different ordering than the default one (`NaturalOrder()`) and see if one of them is more effective in practice.  
I’ve prepared a small script for you.  
You just need to change the variable `order`.

```julia
using SparseMatrixColorings

## Coloring for Jacobians

S = ...  # SparseMatrixCSC
problem = ColoringProblem(; structure=:nonsymmetric, partition=:column)
order = NaturalOrder()
algo = GreedyColoringAlgorithm(order; decompression=:direct)
result = coloring(S, problem, algo)
num_colors = ncolors(result)

## Coloring for Hessians

S = ...  # SparseMatrixCSC
problem = ColoringProblem(; structure=:symmetric, partition=:column)
order = NaturalOrder()
algo = GreedyColoringAlgorithm(order; decompression=:substitution, postprocessing=true)
result = coloring(S, problem, algo)
num_colors = ncolors(result)

## Available ordering

order = NaturalOrder()
order = LargestFirst()
order = SmallestLast()  # <-- very good in practice
order = IncidenceDegree()
order = DynamicLargestFirst()
order = RandomOrder()  # <-- not recommended
```

**Reply by PierreMartinon** [reply]
Hi @amontoison 

Thanks, I'll look into this !
https://github.com/control-toolbox/CTDirect.jl/issues/385

**Comment by ocots** [comment]
@amontoison Thanks!

Just a message: we are organizing the next JuliaCon Local, in Paris at the beginning of October. The [call of proposals](https://juliacon.org/local/paris2025/cfp/) is opened until May 15th. Do not hesitate to spread the word.
